libs/vkd3d/device.c

```diff
@@ -3888,19 +3897,78 @@ static HRESULT STDMETHODCALLTYPE d3d12_device_CreatePipelineState(d3d12_device_i
 static HRESULT STDMETHODCALLTYPE d3d12_device_OpenExistingHeapFromAddress(d3d12_device_iface *iface,
         void *address, REFIID riid, void **heap)
 {
-    FIXME("iface %p, address %p, riid %s, heap %p stub!\n",
-            iface, address, debugstr_guid(riid), heap);
+#ifdef _WIN32
+    MEMORY_BASIC_INFORMATION info;
+    struct d3d12_device *device;
+    struct d3d12_heap *object;
+    size_t allocation_size;
+    HRESULT hr;
 
+    TRACE("iface %p, address %p, riid %s, heap %p\n",
+          iface, address, debugstr_guid(riid), heap);
+
+    if (!VirtualQuery(address, &info, sizeof(info)))
+    {
+        ERR("Failed to VirtualQuery host pointer.\n");
+        return E_INVALIDARG;
+    }
+
+    /* Allocation base must equal address. */
+    if (info.AllocationBase != address)
+        return E_INVALIDARG;
+    if (info.BaseAddress != info.AllocationBase)
+        return E_INVALIDARG;
+
+    /* All pages must be committed. */
+    if (info.State != MEM_COMMIT)
+        return E_INVALIDARG;
+
+    /* We can only have one region of page protection types.
+     * Verify this by querying the end of the range. */
+    allocation_size = info.RegionSize;
+    if (VirtualQuery((uint8_t *)address + allocation_size, &info, sizeof(info)) &&
+            info.AllocationBase == address)
+    {
+        /* All pages must have same protections, so there cannot be multiple regions for VirtualQuery. */
+        return E_INVALIDARG;
+    }
+
+    device = impl_from_ID3D12Device(iface);
+
+    if (FAILED(hr = d3d12_heap_create_from_host_pointer(device, address, allocation_size, &object)))
+    {
+        *heap = NULL;
+        return hr;
+    }
+
+    return return_interface(&object->ID3D12Heap_iface, &IID_ID3D12Heap, riid, heap);
+#else
+    FIXME("OpenExistingHeapFromAddress can only be implemented in native Win32.\n");
     return E_NOTIMPL;
+#endif
 }
 
 static HRESULT STDMETHODCALLTYPE d3d12_device_OpenExistingHeapFromFileMapping(d3d12_device_iface *iface,
         HANDLE file_mapping, REFIID riid, void **heap)
 {
-    FIXME("iface %p, file_mapping %p, riid %s, heap %p stub!\n",
-            iface, file_mapping, debugstr_guid(riid), heap);
+#ifdef _WIN32
+    void *addr;
+    HRESULT hr;
+    TRACE("iface %p, file_mapping %p, riid %s, heap %p\n",
+          iface, file_mapping, debugstr_guid(riid), heap);
+
+    /* 0 size maps everything. */
+    addr = MapViewOfFile(file_mapping, FILE_MAP_ALL_ACCESS, 0, 0, 0);
+    if (!addr)
+        return E_INVALIDARG;
 
+    hr = d3d12_device_OpenExistingHeapFromAddress(iface, addr, riid, heap);
+    UnmapViewOfFile(addr);
```

Comment:
Is it safe to unmap here? Wouldn't this effectively free the virtual address range?

Reply:
It should be, the physical allocation should be ref-counted in the OS.

Reply:
Changed it to unmap at the end to be 100% sure about behavior.